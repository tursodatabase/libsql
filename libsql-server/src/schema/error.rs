use axum::response::IntoResponse;
use hyper::StatusCode;

use crate::error::ResponseError;

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("failed to register migration job: {0}")]
    Registration(Box<dyn std::error::Error + Send + Sync + 'static>),
    #[error("migration scheduler exited")]
    SchedulerExited,
    #[error("corrupted job status: {0}")]
    CorruptedJobStatus(serde_json::Error),
    #[error("sqlite error: {0}")]
    Sqlite(#[from] rusqlite::Error),
}

impl ResponseError for Error {}

impl IntoResponse for &Error {
    fn into_response(self) -> axum::response::Response {
        self.format_err(StatusCode::INTERNAL_SERVER_ERROR)
    }
}
