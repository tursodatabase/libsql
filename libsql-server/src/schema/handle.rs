use std::sync::Arc;

use tokio::sync::{broadcast::Receiver, mpsc, oneshot};

use crate::connection::program::Program;

use super::{error::Error, MigrationJobStatus, SchedulerMessage};

#[derive(Clone)]
pub struct SchedulerHandle {
    sender: mpsc::Sender<SchedulerMessage>,
}

impl From<mpsc::Sender<SchedulerMessage>> for SchedulerHandle {
    fn from(sender: mpsc::Sender<SchedulerMessage>) -> Self {
        Self { sender }
    }
}

pub struct JobHandle {
    job_id: i64,
    notifier: Receiver<(i64, MigrationJobStatus)>,
}

impl JobHandle {
    pub(crate) fn new(job_id: i64, notifier: Receiver<(i64, MigrationJobStatus)>) -> JobHandle {
        Self { job_id, notifier }
    }

    pub async fn wait_for(&mut self, f: impl Fn(MigrationJobStatus) -> bool) {
        while let Ok(next) = self.notifier.recv().await {
            if next.0 == self.job_id && f(next.1) {
                return;
            }
        }
    }

    pub(crate) fn job_id(&self) -> i64 {
        self.job_id
    }
}

impl SchedulerHandle {
    pub(crate) async fn register_migration_task(
        &self,
        schema: crate::namespace::NamespaceName,
        migration: Arc<Program>,
    ) -> Result<JobHandle, Error> {
        let (ret, rcv) = oneshot::channel();
        let msg = SchedulerMessage::ScheduleMigration {
            schema,
            migration,
            ret,
        };
        self.sender
            .send(msg)
            .await
            .map_err(|_| Error::SchedulerExited)?;
        rcv.await.unwrap()
    }

    pub(crate) async fn get_job_status(
        &self,
        job_id: i64,
    ) -> Result<(MigrationJobStatus, Option<String>), Error> {
        let (ret, rcv) = oneshot::channel();
        let msg = SchedulerMessage::GetJobStatus { job_id, ret };
        self.sender
            .send(msg)
            .await
            .map_err(|_| Error::SchedulerExited)?;
        rcv.await.unwrap()
    }
}
