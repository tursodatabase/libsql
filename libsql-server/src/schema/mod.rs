//! This module contains code related to schema migration.
//! When an query that performs writes is sent to a schema database (`shared_schema: true`), then
//! this query is executed exactly one on every database that uses that schema database as a source
//! of truth (`shared_schema_name: '<my_schema_db>`), by way of this schema migration process:
//! - The migration is first performed against the schema db in a dry run (we create a transaction,
//! execute the migration and rollback and roll it back).
//! - If the schema db migration was successfull then we enqueue a `MigrationJob` in the job queue.
//! Also, for every database referring to the schema, we enqueue a `MigrationTask` that refers to
//! the job.
//! - The `Scheduler` is in charge of stepping jobs and tasks to completion.
//! - In the initial state of the migration, all the tasks are in `MigrationTaskStatus::Enqueue` state,
//! and the corresponding job is in `MigrationJobStatus::WaitingForDryRun`.
//! - The scheduler gets a batch of task in the `Enqueued` state from the database, and schedules
//! them on for a dry-run on the its worker pool.
//! - When a task is in `Enqueued` state, it is first enqueued to the database's own queue (a table
//! named `__libsql_migration_tasks`), and it is set to `Enqueued`. The scheduler then performs a
//! dry run on the migration on that database, setting the outcome of the dry-run to the databases
//! queue in the same transaction. If the dry-run was successfull, the status is set to
//! `DryRunSuccessfull`, otherwise, it is set to `DryRunFailure`, and the error message is saved.
//! By atomically changing the status of the task as we perform the migration, we ensure that in
//! case of failure, we'll always either safely re-perform the task, or collect the result of a
//! prior execution.
//! - The scheduler attempts to drive all tasks to the state `DryRunSuccessfull` before stepping
//! the job's status. If one tasks fails the dry-run, then the migration is aborted.
//! - If all the tasks sucessfully performed the dry-run, then the job is stepped to
//! `DryRunSuccessfull`, at which point, potential waiter are notified.
//! - The scheduler then steps the job to the `WaitingRun` status, and the task status is updated
//! to `Run`. Each task now perform the migration for real, and report their status, in the same
//! manner as the dry-run, except for the fact that the migration is actually committed.
//! - If all tasks are successfull, then the scheduler performs the migration on the schema, and
//! update the job's state to it's final state, `RunSuccess`.
pub(crate) mod db;
mod error;
mod handle;
mod message;
mod migration;
mod result_builder;
mod scheduler;
mod status;

pub use db::{get_migration_details, get_migrations_summary};
pub use error::Error;
pub use handle::SchedulerHandle;
pub use message::SchedulerMessage;
pub use migration::*;
pub use scheduler::Scheduler;
pub use status::{MigrationDetails, MigrationJobStatus, MigrationSummary, MigrationTaskStatus};

use crate::connection::program::Program;

pub fn validate_migration(migration: &Program) -> Result<(), Error> {
    if migration.steps().iter().any(|s| s.query.stmt.kind.is_txn()) {
        Err(Error::MigrationContainsTransactionStatements)
    } else {
        Ok(())
    }
}
