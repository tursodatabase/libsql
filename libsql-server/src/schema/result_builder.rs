use std::time::Duration;

use crate::query_result_builder::QueryResultBuilder;

pub struct SchemaMigrationResultBuilder<B> {
    inner: B,
    step: usize,
    errors: Vec<(usize, String)>,
}

impl<B> SchemaMigrationResultBuilder<B> {
    pub(crate) fn new(inner: B) -> Self {
        Self {
            inner,
            step: 0,
            errors: Vec::new(),
        }
    }

    pub(crate) fn into_inner(self) -> B {
        self.inner
    }

    pub(crate) fn is_success(&self) -> bool {
        self.errors.is_empty()
    }

    pub(crate) fn into_error(self) -> (usize, String) {
        assert!(!self.errors.is_empty());
        self.errors.into_iter().next().unwrap()
    }
}

impl<B: QueryResultBuilder> QueryResultBuilder for SchemaMigrationResultBuilder<B> {
    type Ret = B::Ret;

    fn init(
        &mut self,
        config: &crate::query_result_builder::QueryBuilderConfig,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.step = 0;
        self.inner.init(config)
    }

    fn begin_step(&mut self) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.step += 1;
        self.inner.begin_step()
    }

    fn finish_step(
        &mut self,
        affected_row_count: u64,
        last_insert_rowid: Option<i64>,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner
            .finish_step(affected_row_count, last_insert_rowid)
    }

    fn step_error(
        &mut self,
        error: crate::error::Error,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.errors.push((self.step, error.to_string()));
        self.inner.step_error(error)
    }

    fn cols_description<'a>(
        &mut self,
        cols: impl IntoIterator<Item = impl Into<crate::query_result_builder::Column<'a>>>,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.cols_description(cols)
    }

    fn begin_rows(&mut self) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.begin_rows()
    }

    fn begin_row(&mut self) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.begin_row()
    }

    fn add_row_value(
        &mut self,
        v: rusqlite::types::ValueRef,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.add_row_value(v)
    }

    fn finish_row(&mut self) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.finish_row()
    }

    fn finish_rows(&mut self) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.finish_rows()
    }

    fn finish(
        &mut self,
        last_frame_no: Option<crate::replication::FrameNo>,
        is_auto_commit: bool,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.finish(last_frame_no, is_auto_commit)
    }

    fn into_ret(self) -> Self::Ret {
        self.inner.into_ret()
    }

    fn add_stats(&mut self, rows_read: u64, rows_written: u64, duration: Duration) {
        self.inner.add_stats(rows_read, rows_written, duration);
    }
}
