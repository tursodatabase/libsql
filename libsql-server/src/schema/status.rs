use serde::{Deserialize, Serialize};

/// Represents the status of a migration task
#[derive(Debug, Serialize, Deserialize, Copy, Clone)]
#[repr(u64)]
pub enum MigrationTaskStatus {
    /// The task was enqueued, and shoudl perform a dry run
    Enqueued = 0,
    /// The dry run was successfull
    DryRunSuccess = 1,
    /// The dry run failed
    DryRunFailure = 2,
    /// The migration task should be performed
    Run = 3,
    /// The migration task was a success
    Success = 4,
    /// The migration task was a failure
    Failure = 5,
}

impl MigrationTaskStatus {
    pub fn from_int(i: u64) -> Self {
        match i {
            0 => Self::Enqueued,
            1 => Self::DryRunSuccess,
            2 => Self::DryRunFailure,
            3 => Self::Run,
            4 => Self::Success,
            5 => Self::Failure,
            _ => panic!(),
        }
    }

    pub fn is_failure(&self) -> bool {
        matches!(self, Self::DryRunFailure | Self::Failure)
    }

    pub const fn num_variants() -> usize {
        // the only use of this is to create a compile error if someone adds a variant
        match Self::Enqueued {
            MigrationTaskStatus::Enqueued => (),
            MigrationTaskStatus::DryRunSuccess => (),
            MigrationTaskStatus::DryRunFailure => (),
            MigrationTaskStatus::Run => (),
            MigrationTaskStatus::Success => (),
            MigrationTaskStatus::Failure => (),
        }

        // don't forget to update that!
        6
    }
}

/// Represents the status of a migration job
#[derive(Debug, Serialize, Deserialize, Copy, Clone, PartialEq, Eq)]
#[repr(u64)]
pub enum MigrationJobStatus {
    /// Waiting for all tasks to report the dry run status
    WaitingDryRun = 0,
    /// All tasks reported a successfull dry run status
    DryRunSuccess = 1,
    /// One or more task reported an unsuccessful dry drun
    DryRunFailure = 2,
    /// Waiting for all the tasks to return a successful migration run
    WaitingRun = 3,
    /// All tasks retuned successfully, and teh schema was updated successfully
    RunSuccess = 4,
    /// something fucked up
    RunFailure = 5,
}

impl MigrationJobStatus {
    pub fn from_int(i: u64) -> Self {
        match i {
            0 => Self::WaitingDryRun,
            1 => Self::DryRunSuccess,
            2 => Self::DryRunFailure,
            3 => Self::WaitingRun,
            4 => Self::RunSuccess,
            5 => Self::RunFailure,
            _ => panic!(),
        }
    }

    /// Returns `true` if the migration job status is [`WaitingRun`].
    ///
    /// [`WaitingRun`]: MigrationJobStatus::WaitingRun
    #[must_use]
    pub fn is_waiting_run(&self) -> bool {
        matches!(self, Self::WaitingRun)
    }
}
