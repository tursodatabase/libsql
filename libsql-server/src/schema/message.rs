use std::sync::Arc;

use tokio::sync::oneshot;

use crate::connection::program::Program;
use crate::namespace::NamespaceName;

use super::error::Error;
use super::handle::JobHandle;
use super::MigrationJobStatus;

pub enum SchedulerMessage {
    ScheduleMigration {
        schema: NamespaceName,
        migration: Arc<Program>,
        ret: oneshot::Sender<Result<JobHandle, Error>>,
    },
    GetJobStatus {
        job_id: i64,
        ret: oneshot::Sender<Result<(MigrationJobStatus, Option<String>), Error>>,
    },
}
