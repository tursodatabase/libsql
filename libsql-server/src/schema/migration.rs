use itertools::Itertools;
use once_cell::sync::Lazy;
use rusqlite::Savepoint;

use crate::connection::program::{Program, Vm};
use crate::query_result_builder::{IgnoreResult, QueryBuilderConfig, QueryResultBuilder};

use super::result_builder::SchemaMigrationResultBuilder;
use super::status::MigrationTask;
use super::{Error, MigrationTaskStatus};

pub fn setup_migration_table(conn: &mut rusqlite::Connection) -> Result<(), Error> {
    static TASKS_TABLE_QUERY: Lazy<String> = Lazy::new(|| {
        format!(
            "CREATE TABLE IF NOT EXISTS __libsql_migration_tasks (
                job_id INTEGER PRIMARY KEY,
                status INTEGER,
                migration TEXT NOT NULL,
                error TEXT,
                finished BOOLEAN GENERATED ALWAYS AS ({})
            )",
            MigrationTaskStatus::finished_states()
                .iter()
                .map(|s| format!("status = {}", *s as u64))
                .join(" OR ")
        )
    });

    let tx = conn.transaction()?;
    let schema_version =
        tx.pragma_query_value(None, "schema_version", |row| Ok(row.get::<_, u64>(0)?))?;
    tx.execute(&*TASKS_TABLE_QUERY, ())?;
    // We have to make sure schema_version is not changed in case this db is a fork of shared schema
    // or a namespace that links to shared schema.
    // It is ok to do this because we're the only connection to that database at this point.
    tx.pragma_update(None, "schema_version", schema_version)?;
    tx.commit()?;
    Ok(())
}

pub fn has_pending_migration_task(conn: &rusqlite::Connection) -> Result<bool, Error> {
    Ok(conn.query_row(
        "SELECT COUNT(1) FROM __libsql_migration_tasks WHERE finished = false",
        (),
        |row| Ok(row.get::<_, usize>(0)? != 0),
    )?)
}

pub fn enqueue_migration_task(
    conn: &rusqlite::Connection,
    task: &MigrationTask,
    migration: &Program,
) -> Result<(), Error> {
    let migration = serde_json::to_string(migration).unwrap();
    conn.execute(
        "INSERT OR IGNORE INTO __libsql_migration_tasks (job_id, status, migration) VALUES (?, ?, ?)",
        (task.job_id(), *task.status() as u64, &migration),
    )?;

    Ok(())
}

pub fn abort_migration_task(
    conn: &rusqlite::Connection,
    task: &MigrationTask,
) -> Result<(), Error> {
    // there is a `NOT NULL` constraint on migration, but if we are aborting a task that wasn't
    // already enqueued, we need a placeholder. It's ok because we are never gonna try to run a
    // failed task migration.
    conn.execute("INSERT OR REPLACE INTO __libsql_migration_tasks (job_id, status, error, migration) VALUES (?, ?, ?, ?)",
    (task.job_id, MigrationTaskStatus::Failure as u64, "aborted", "aborted"))?;

    Ok(())
}

/// set the task status to `Run` if its current state is `DryRunSuccess`
pub fn step_migration_task_run(conn: &rusqlite::Connection, job_id: i64) -> Result<(), Error> {
    conn.execute(
        "
            UPDATE __libsql_migration_tasks
            SET status = ?
            WHERE job_id = ? AND status = ?
            ",
        (
            MigrationTaskStatus::Run as u64,
            job_id,
            MigrationTaskStatus::DryRunSuccess as u64,
        ),
    )?;

    Ok(())
}

fn get_task_infos(
    conn: &rusqlite::Connection,
    job_id: i64,
) -> Result<(MigrationTaskStatus, Option<Program>, Option<String>), Error> {
    Ok(conn.query_row(
        "SELECT status, migration, error FROM __libsql_migration_tasks WHERE job_id = ?",
        [job_id],
        |row| {
            let status = MigrationTaskStatus::from_int(row.get::<_, u64>(0)?);
            let (migration, error) = match status {
                MigrationTaskStatus::Enqueued | MigrationTaskStatus::Run => {
                    let migration: Program =
                        serde_json::from_str(row.get_ref(1)?.as_str()?).unwrap();
                    (Some(migration), None)
                }
                MigrationTaskStatus::DryRunSuccess | MigrationTaskStatus::Success => (None, None),
                MigrationTaskStatus::DryRunFailure | MigrationTaskStatus::Failure => {
                    let error = row.get::<_, Option<String>>(2)?;
                    (None, error)
                }
            };

            Ok((status, migration, error))
        },
    )?)
}

pub(super) fn step_task(
    txn: &mut rusqlite::Transaction,
    job_id: i64,
) -> Result<(MigrationTaskStatus, Option<String>), Error> {
    let (current_state, migration, error) = get_task_infos(txn, job_id)?;

    match current_state {
        MigrationTaskStatus::DryRunSuccess | MigrationTaskStatus::DryRunFailure => {
            Ok((current_state, error))
        }
        MigrationTaskStatus::Run | MigrationTaskStatus::Enqueued => {
            let (ret, new_status) = perform_migration(
                txn,
                migration.as_ref().unwrap(),
                current_state.is_enqueued(),
                IgnoreResult,
                &QueryBuilderConfig::default(),
            );
            let error = ret.err().map(|e| e.to_string());
            update_db_task_status(txn, job_id, new_status, error.as_deref())?;

            Ok((new_status, error))
        }
        // final state, nothing to do but report
        MigrationTaskStatus::Success | MigrationTaskStatus::Failure => Ok((current_state, error)),
    }
}

pub fn perform_migration<B: QueryResultBuilder>(
    txn: &mut rusqlite::Transaction,
    migration: &Program,
    dry_run: bool,
    builder: B,
    config: &QueryBuilderConfig,
) -> (Result<B, Error>, MigrationTaskStatus) {
    // todo error handling is sketchy, improve
    let builder = SchemaMigrationResultBuilder::new(builder);
    let mut savepoint = txn.savepoint().unwrap();
    match try_perform_migration(&mut savepoint, migration, builder, config) {
        Ok(builder) if builder.is_success() => {
            let status = if dry_run {
                savepoint.rollback().unwrap();
                drop(savepoint);
                MigrationTaskStatus::DryRunSuccess
            } else {
                savepoint.commit().unwrap();
                MigrationTaskStatus::Success
            };
            (Ok(builder.into_inner()), status)
        }
        Ok(builder) => {
            assert!(!builder.is_success());
            savepoint.rollback().unwrap();
            drop(savepoint);
            let status = if dry_run {
                MigrationTaskStatus::DryRunFailure
            } else {
                MigrationTaskStatus::Failure
            };
            let (step, error) = builder.into_error();
            (Err(Error::MigrationError(step, error)), status)
        }
        Err(e) => {
            savepoint.rollback().unwrap();
            drop(savepoint);
            let status = if dry_run {
                MigrationTaskStatus::DryRunFailure
            } else {
                MigrationTaskStatus::Failure
            };
            (Err(e), status)
        }
    }
}

pub(super) fn update_db_task_status(
    conn: &rusqlite::Connection,
    job_id: i64,
    status: MigrationTaskStatus,
    error: Option<&str>,
) -> Result<(), Error> {
    conn.execute(
        "UPDATE __libsql_migration_tasks SET status = ?, error = ? WHERE job_id = ?",
        (status as u64, error, job_id),
    )?;

    Ok(())
}

fn try_perform_migration<B: QueryResultBuilder>(
    savepoint: &mut Savepoint,
    migration: &Program,
    mut builder: B,
    config: &QueryBuilderConfig,
) -> Result<B, Error> {
    builder.init(config).unwrap();
    let mut vm = Vm::new(
        builder, // todo use proper builder
        migration,
        |_| (false, None),
        |_, _, _| (),
    );

    while !vm.finished() {
        vm.step(savepoint).unwrap(); // return migration error
    }

    vm.builder().finish(None, true).unwrap();

    Ok(vm.into_builder())
}

#[cfg(test)]
mod test {
    use insta::assert_debug_snapshot;
    use libsql_sys::wal::Sqlite3WalManager;
    use tempfile::tempdir;

    use crate::connection::libsql::open_conn_active_checkpoint;
    use crate::namespace::NamespaceName;

    use super::*;

    #[test]
    fn already_performed_task_is_not_reexecuted() {
        let tmp = tempdir().unwrap();

        let mut conn =
            open_conn_active_checkpoint(tmp.path(), Sqlite3WalManager::default(), None, 1000, None)
                .unwrap();
        setup_migration_table(&mut conn).unwrap();

        let task = MigrationTask {
            namespace: NamespaceName::default(),
            status: MigrationTaskStatus::Success,
            job_id: 1,
            task_id: 1,
        };
        enqueue_migration_task(&conn, &task, &Program::seq(&["create table test (x)"])).unwrap();
        let mut txn = conn.transaction().unwrap();
        let (status, error) = step_task(&mut txn, 1).unwrap();
        txn.commit().unwrap();

        assert!(error.is_none());
        assert_eq!(status, MigrationTaskStatus::Success);

        // The migration should not have been performed, since we declared that it was already
        // successfully executed.
        let schema = conn
            .prepare("select * from sqlite_schema")
            .unwrap()
            .query_map((), |r| Ok(format!("{r:?}")))
            .unwrap()
            .collect::<Vec<_>>();
        assert_debug_snapshot!(schema);
    }

    #[test]
    fn ignore_task_enqueue_if_already_exists() {
        let tmp = tempdir().unwrap();

        let mut conn =
            open_conn_active_checkpoint(tmp.path(), Sqlite3WalManager::default(), None, 1000, None)
                .unwrap();
        setup_migration_table(&mut conn).unwrap();

        let task = MigrationTask {
            namespace: NamespaceName::default(),
            status: MigrationTaskStatus::Success,
            job_id: 1,
            task_id: 1,
        };
        enqueue_migration_task(&conn, &task, &Program::seq(&["create table test (x)"])).unwrap();

        let task = MigrationTask {
            namespace: NamespaceName::default(),
            status: MigrationTaskStatus::Enqueued,
            job_id: 1,
            task_id: 1,
        };
        enqueue_migration_task(&conn, &task, &Program::seq(&["create table test (x)"])).unwrap();
        let (status, _, _) = get_task_infos(&conn, 1).unwrap();
        assert_eq!(status, MigrationTaskStatus::Success);
    }
}
