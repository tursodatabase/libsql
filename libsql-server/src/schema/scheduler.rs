use std::sync::Arc;

use tokio::sync::{mpsc, OwnedSemaphorePermit, Semaphore};
use tokio::task::JoinSet;

use crate::connection::program::Program;
use crate::connection::MakeConnection;
use crate::namespace::meta_store::{MigrationJob, MigrationTask};
use crate::namespace::{NamespaceName, NamespaceStore};
use crate::query_result_builder::IgnoreResult;
use crate::schema::{step_migration_task_run, MigrationJobStatus};

use super::error::Error;
use super::migration::enqueue_migration_task;
use super::{perform_migration, MigrationTaskStatus, SchedulerMessage};

enum WorkResult {
    Task {
        old_status: MigrationTaskStatus,
        task: MigrationTask,
        error: Option<String>,
    },
    Schema {
        job_id: i64,
    },
}

pub struct Scheduler {
    namespace_store: NamespaceStore,
    workers: JoinSet<WorkResult>,
    current_batch: Vec<MigrationTask>,
    // job id of the curently processing job
    current_job: Option<MigrationJob>,
    has_work: bool,
}

impl Scheduler {
    pub async fn run(mut self, mut receiver: mpsc::Receiver<SchedulerMessage>) {
        let tasks_permits = Arc::new(Semaphore::new(10));
        loop {
            tokio::select! {
                Some(msg) = receiver.recv() => {
                    self.handle_msg(msg).await;
                }
                // There is work to do, and a worker slot to perform it
                Ok(permit) = tasks_permits.clone().acquire_owned(), if self.has_work => {
                    self.enqueue_work(permit).await;
                }
                Some(res) = self.workers.join_next(), if !self.workers.is_empty() => {
                    self.has_work = true;
                    match res {
                        Ok(WorkResult::Schema { job_id }) => {
                            let job = self.current_job.take().unwrap();
                            assert_eq!(job.job_id(), job_id);
                            self.namespace_store.meta_store().update_job_status(job_id, MigrationJobStatus::RunSuccess).await.unwrap();
                        }
                        Ok(WorkResult::Task { old_status, task, error }) => {
                            let new_status = *task.status();
                            self.namespace_store
                                .meta_store()
                                .update_task_status(task, error).await.unwrap();
                            let current_job = self.current_job
                                .as_mut()
                                .expect("processing task result, but job is missing");

                            *current_job.progress_mut(old_status) -= 1;
                            *current_job.progress_mut(new_status) += 1;
                        }
                        Err(_e) => {
                            todo!("migration task panicked");
                        }
                    }
                }
                else => break,
            }
        }

        tracing::info!("all scheduler handles dropped: exiting.");
    }

    async fn handle_msg(&mut self, msg: SchedulerMessage) {
        match msg {
            SchedulerMessage::ScheduleMigration {
                schema,
                migration,
                ret,
            } => {
                let res = self.register_migration_task(schema, migration).await;
                let _ = ret.send(res);
                self.has_work = true;
            }
        }
    }

    // TODO: refactor this function it's turning into a mess. Not so simple, because of borrow
    // constraints
    async fn enqueue_work(&mut self, permit: OwnedSemaphorePermit) {
        let job = match self.current_job {
            Some(ref mut job) => job,
            None => {
                match self
                    .namespace_store
                    .meta_store()
                    .get_next_pending_job()
                    .await
                    .unwrap()
                {
                    Some(job) => self.current_job.insert(job),
                    None => {
                        self.has_work = false;
                        return;
                    }
                }
            }
        };

        // try to step the current job
        match *job.status() {
            MigrationJobStatus::WaitingDryRun => {
                // there was a dry run failure, abort the task
                if job.progress(MigrationTaskStatus::DryRunFailure) != 0 {
                    todo!("abort job");
                }
                if job.progress_all(MigrationTaskStatus::DryRunSuccess) {
                    // ready to move to the run phase
                    // report dry run succes
                    // todo: it may be worthwhile to check that all the db state reflects our
                    // vision of the tasks progress here
                    // also if we wanted to abort before the run, now would be the right place

                    *job = self
                        .namespace_store
                        .meta_store()
                        .job_step_dry_run_success(job.clone())
                        .await
                        .unwrap();
                    if matches!(job.status(), MigrationJobStatus::DryRunSuccess) {
                        // todo!("notify dry run success")
                    }
                }
            }
            MigrationJobStatus::DryRunSuccess => {
                self.namespace_store
                    .meta_store()
                    .update_job_status(job.job_id(), MigrationJobStatus::WaitingRun)
                    .await
                    .unwrap();
                *job.status_mut() = MigrationJobStatus::WaitingRun;
            }
            MigrationJobStatus::DryRunFailure => todo!(),
            MigrationJobStatus::WaitingRun => {
                // there was a dry run failure, abort the task
                if job.progress(MigrationTaskStatus::Failure) != 0 {
                    todo!("that shouldn't happen!");
                }
                if job.progress_all(MigrationTaskStatus::Success) {
                    // todo: perform more robust check, like in waiting dry run

                    let connection_maker = self
                        .namespace_store
                        .with(job.schema(), |ns| {
                            ns.db
                                .as_schema()
                                .expect("expected database to be a schema database")
                                .connection_maker()
                                .clone()
                        })
                        .await
                        .unwrap();

                    let connection = connection_maker.create().await.unwrap();
                    // todo: make sure that the next job is not a job for the same namesapce:
                    // prevent job to be enqueue for a schema if there is still onging work for
                    // that schema
                    let job_id = job.job_id();
                    let migration = job.migration();
                    self.workers.spawn_blocking(move || {
                        let _perm = permit;
                        connection.connection().with_raw(|conn| {
                            let mut txn = conn.transaction().unwrap();
                            let schema_version = txn
                                .query_row("PRAGMA schema_version", (), |row| row.get::<_, i64>(0))
                                .unwrap();

                            if schema_version != job_id {
                                // todo: use proper builder and collect errors
                                let (ret, _status) =
                                    perform_migration(&mut txn, &migration, false, IgnoreResult);
                                let _error = ret.err().map(|e| e.to_string());
                                txn.pragma_update(None, "schema_version", job_id).unwrap();
                                // update schema version to job_id?
                                txn.commit().unwrap();
                            }

                            WorkResult::Schema { job_id }
                        })
                    });
                    // do not enqueue anything until the schema migration is complete
                    self.has_work = false;
                    return;
                }
            }
            MigrationJobStatus::RunSuccess => unreachable!(),
            MigrationJobStatus::RunFailure => todo!("handle run failure"),
        }

        // fill the current batch if necessary
        if self.current_batch.is_empty() {
            match job.status() {
                MigrationJobStatus::WaitingDryRun => {
                    // get a batch of enqueued tasks
                    self.current_batch = self
                        .namespace_store
                        .meta_store()
                        .get_next_pending_migration_tasks_batch(
                            job.job_id(),
                            MigrationTaskStatus::Enqueued,
                            50, // TODO: make that configurable maybe?
                        )
                        .await
                        .unwrap();
                }
                MigrationJobStatus::WaitingRun => {
                    // get a batch of enqueued tasks
                    self.current_batch = self
                        .namespace_store
                        .meta_store()
                        .get_next_pending_migration_tasks_batch(
                            job.job_id(),
                            MigrationTaskStatus::DryRunSuccess,
                            50, // TODO: make that configurable maybe?
                        )
                        .await
                        .unwrap();
                }
                _ => (),
            };
        }

        // enqueue some work
        if let Some(mut task) = self.current_batch.pop() {
            let connection_maker = self
                .namespace_store
                .with(task.namespace(), |ns| {
                    ns.db
                        .as_primary()
                        .expect("attempting to perform schema migration on non-primary database")
                        .connection_maker()
                        .clone()
                })
                .await
                .unwrap();

            let connection = connection_maker.create().await.unwrap();

            let migration = job.migration();
            let job_status = *job.status();
            self.workers.spawn_blocking(move || {
                let old_status = *task.status();
                // move the permit inside of the closure, so that it gets dropped when the work is done.
                let _permit = permit;
                connection.with_raw(move |conn| {
                    let mut txn = conn.transaction().unwrap();

                    let is_dry_run = match task.status() {
                        MigrationTaskStatus::Enqueued => {
                            enqueue_migration_task(&txn, &task, &migration).unwrap();
                            true
                        }
                        MigrationTaskStatus::DryRunSuccess if job_status.is_waiting_run() => {
                            step_migration_task_run(&txn, &task).unwrap();
                            false
                        }
                        _ => unreachable!("expected task status to be `enqueued` or `run`"),
                    };
                    let (ret, status) =
                        perform_migration(&mut txn, &migration, is_dry_run, IgnoreResult);
                    let error = ret.err().map(|e| e.to_string());
                    super::migration::update_task_status(
                        &txn,
                        task.job_id(),
                        status,
                        error.as_deref(),
                    )
                    .unwrap();
                    *task.status_mut() = status;
                    txn.commit().unwrap();
                    WorkResult::Task {
                        old_status,
                        task,
                        error,
                    }
                })
            });
        } else {
            // there is still job, but the queue is empty, it means that we are waiting for the
            // remaining jobs to report status. just wait.
            self.has_work = false;
        }
    }

    pub async fn register_migration_task(
        &self,
        schema: NamespaceName,
        migration: Arc<Program>,
    ) -> Result<i64, Error> {
        let job_id = self
            .namespace_store
            .meta_store()
            .register_schema_migration(schema, migration)
            .await
            .map_err(|e| Error::Registration(Box::new(e)))?;
        Ok(job_id)
    }

    pub(crate) fn new(namespace_store: NamespaceStore) -> Self {
        Self {
            namespace_store,
            workers: Default::default(),
            current_batch: Vec::new(),
            current_job: None,
            // initialized to true to kickoff the queue
            has_work: true,
        }
    }
}
