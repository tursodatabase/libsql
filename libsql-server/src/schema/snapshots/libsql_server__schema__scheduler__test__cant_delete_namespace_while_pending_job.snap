---
source: libsql-server/src/schema/scheduler.rs
expression: "store.destroy(\"ns\".into()).await.unwrap_err()"
---
PendingMigrationOnSchema(
    schema,
)
